File: src/shared/fetcher.ts
import axios, { AxiosInstance, AxiosError } from 'axios';
import logger from './logger.js';

interface FetcherOptions {
  timeout?: number;
  maxRetries?: number;
  delayMs?: number;
}

const DEFAULT_TIMEOUT = 10_000;
const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_DELAY_MS = 500;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function isRetryable(error: AxiosError): boolean {
  if (error.code === 'ECONNABORTED' || error.code === 'ETIMEDOUT') return true;
  const status = error.response?.status;
  if (status && status >= 500) return true;
  return false;
}

export class Fetcher {
  private client: AxiosInstance;
  private maxRetries: number;
  private delayMs: number;
  private lastRequestTime = 0;

  constructor(options: FetcherOptions = {}) {
    this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
    this.delayMs = options.delayMs ?? DEFAULT_DELAY_MS;
    this.client = axios.create({
      timeout: options.timeout ?? DEFAULT_TIMEOUT,
      headers: {
        'User-Agent':
          'Mozilla/5.0 (compatible; SKKUverseCrawler/1.0)',
      },
      responseType: 'text',
    });
  }

  async fetchBinary(url: string): Promise<Buffer> {
    const now = Date.now();
    const elapsed = now - this.lastRequestTime;
    if (elapsed < this.delayMs) {
      await sleep(this.delayMs - elapsed);
    }

    let lastError: Error | null = null;

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        this.lastRequestTime = Date.now();
        const response = await this.client.get(url, { responseType: 'arraybuffer' });
        return Buffer.from(response.data);
      } catch (err) {
        const axiosErr = err as AxiosError;
        lastError = axiosErr;

        if (!isRetryable(axiosErr)) {
          logger.warn({ url, status: axiosErr.response?.status }, 'Non-retryable fetch error');
          throw axiosErr;
        }

        if (attempt < this.maxRetries) {
          const backoff = Math.pow(2, attempt - 1) * 1000;
          logger.warn(
            { url, attempt, backoff, code: axiosErr.code },
            'Retrying fetch'
          );
          await sleep(backoff);
        }
      }
    }

    logger.error({ url }, 'All fetch retries exhausted');
    throw lastError!;
  }

  async fetch(url: string): Promise<string> {
    const now = Date.now();
    const elapsed = now - this.lastRequestTime;
    if (elapsed < this.delayMs) {
      await sleep(this.delayMs - elapsed);
    }

    let lastError: Error | null = null;

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        this.lastRequestTime = Date.now();
        const response = await this.client.get<string>(url);
        return response.data;
      } catch (err) {
        const axiosErr = err as AxiosError;
        lastError = axiosErr;

        if (!isRetryable(axiosErr)) {
          logger.warn({ url, status: axiosErr.response?.status }, 'Non-retryable fetch error');
          throw axiosErr;
        }

        if (attempt < this.maxRetries) {
          const backoff = Math.pow(2, attempt - 1) * 1000;
          logger.warn(
            { url, attempt, backoff, code: axiosErr.code },
            'Retrying fetch'
          );
          await sleep(backoff);
        }
      }
    }

    logger.error({ url }, 'All fetch retries exhausted');
    throw lastError!;
  }
}

File: py/src/skkuverse_crawler/shared/fetcher.py
from __future__ import annotations

import asyncio
import time

import httpx

from .logger import get_logger

logger = get_logger("fetcher")

DEFAULT_TIMEOUT = 10.0
DEFAULT_MAX_RETRIES = 3
DEFAULT_DELAY_MS = 500


def _is_retryable(exc: BaseException) -> bool:
    if isinstance(exc, httpx.TimeoutException):
        return True
    if isinstance(exc, httpx.HTTPStatusError):
        return exc.response.status_code >= 500
    return False


class Fetcher:
    def __init__(
        self,
        *,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
        delay_ms: int = DEFAULT_DELAY_MS,
    ) -> None:
        self.max_retries = max_retries
        self.delay_ms = delay_ms
        self._last_request_time: float = 0.0
        self._client = httpx.AsyncClient(
            timeout=timeout,
            headers={"User-Agent": "Mozilla/5.0 (compatible; SKKUverseCrawler/1.0)"},
            follow_redirects=True,
        )

    async def _rate_limit(self) -> None:
        now = time.monotonic()
        elapsed_ms = (now - self._last_request_time) * 1000
        if elapsed_ms < self.delay_ms:
            await asyncio.sleep((self.delay_ms - elapsed_ms) / 1000)

    async def _fetch_with_retry(self, url: str) -> httpx.Response:
        await self._rate_limit()

        last_error: Exception | None = None
        for attempt in range(1, self.max_retries + 1):
            try:
                self._last_request_time = time.monotonic()
                resp = await self._client.get(url)
                resp.raise_for_status()
                return resp
            except Exception as exc:
                last_error = exc
                if not _is_retryable(exc):
                    logger.warning("non_retryable_error", url=url, error=str(exc))
                    raise
                if attempt < self.max_retries:
                    backoff = (2 ** (attempt - 1)) * 1.0
                    logger.warning(
                        "retrying_fetch", url=url, attempt=attempt, backoff=backoff
                    )
                    await asyncio.sleep(backoff)

        logger.error("all_retries_exhausted", url=url)
        raise last_error  # type: ignore[misc]

    async def fetch(self, url: str) -> str:
        return (await self._fetch_with_retry(url)).text

    async def fetch_binary(self, url: str) -> bytes:
        return (await self._fetch_with_retry(url)).content

    async def close(self) -> None:
        await self._client.aclose()
